### ISSUE TITLE
Table `user_group_map` is altered unexpectedly - preventing access to some units

### ISSUE THREAD

**Issue by franck-boullier** [unee-t/bz-database#90]
# The problem:

For (yet) unknown reasons, in the DEV environment, the table `user_group_map` is altered unexpectedly

This happened again on 25/10/2018 - 01h27 despite the fix implemented in PR #89

**Comment by kaihendry** [comment]
Isn't this the same as https://github.com/unee-t/bz-database/issues/88 ?

**Comment by franck-boullier** [comment]
> Isn't this the same as #88 ?

Not really: we don't know for sure that this is caused by the invitation procedure yet

**Comment by franck-boullier** [comment]
PR #89 should have fixed this but we'll keep that open for a few days in case the issue appears again...

**Comment by kaihendry** [comment]
This is still an issue in dev :( https://media.dev.unee-t.com/2018-11-07/user_group_map.mp4

**Comment by franck-boullier** [comment]
I have no clue on what might be causing that...
Can we look at the log or get the exact time when the table was nuked?
If I have the exact time I might be able to comb the SQL log to understand what might have triggered that... (emphasis on "might") but we need to understand what's happening there: this is a huge risk...

**Comment by franck-boullier** [comment]
See https://media.dev.unee-t.com/2018-11-07/cloudwatch.mp4 for more information on the issue

**Comment by franck-boullier** [comment]
Based on @kaihendry's analysis of the cloudwatch log, we need to take another look at the script 'invite_user_in_a_role_in_a_unit.sql'

**Comment by kaihendry** [comment]
All that script does https://github.com/unee-t/invite/blob/master/sql/invite_user_in_a_role_in_a_unit.sql#L45 is call https://media.dev.unee-t.com/2018-11-08/add_user_to_role_in_unit.txt

**Comment by franck-boullier** [comment]
Pasting a comment taken from closed PR #89

Did some analysis on the mysql bin logs with @kaihendry , following is the filtered logs. Hopefully with the timestamps provided, the actual API call can be narrowed down.

1) TRUNCATE TABLE `user_group_map`, and TRUNCATE TABLE `group_group_map` is called multiple times : attached is the filtered logs with timestamp

[truncate.log](https://github.com/unee-t/bz-database/files/2559914/truncate.log)

2) DELETE logs are also filtered and attached here with timestamp, note nothing suspicious found here by me. However I severely lack understanding of how the internal db structures are suppose to work. So hard for me to say, if issue not found in above, someone might want to review this.

[delete.log](https://github.com/unee-t/bz-database/files/2559915/delete.log)

3) Side note : What is “ut_user_group_map_temp” or “ut_user_group_map_dedup” table for ? - not too sure if there are race conditions involved with them? And if they are related to this issue. But it does look suspicious to me.

Btw : full logs was about 160mb after parsing it from raw form, if anyone wants it - let me know - not posting here, in case there is any confidential information inside (ie. password hashes).

**Comment by franck-boullier** [comment]
Ahah! moment
There seem to be an old procedure `unit_create_with_dummy_users` which does not use the new method to assign user permissions...
exploring that and how we can fix this

**Comment by franck-boullier** [comment]
I am now 90% sure that the issue happens when a user tries to create a new unit.
currently fixing the script `unit_create_with_dummy_users` (it needs to be regardless anyway)

@kaihendry can you help me check if there is a correlation in the log between the the 2 events:
- a user tries to create a new unit
- the table `user_group_map` is truncated

**Comment by franck-boullier** [comment]
I think I'm almost there...
I will stop for the night and resume in a few hours my time.
I will then do the BZ schema update in the DEV environment so we can test again if:
- nothing is broken and we can still:
  - create new units
  - invite new users
- This fixes the issue of the tuncated `user_group_map` table

@kaihendry and @nbiton let me see if you have an issue with that plan

**Comment by franck-boullier** [comment]
it seems that `user_group_map` has been deleted again last night...

I'll be doing the BZ Schema upgrade in a few minutes

@kaihendry can you please:
- Restore the table `user_group_map` from a backup
- Ask UILicious to run the exact same thing that they did last night around the time when the table `user_group_map` as been nuked ---> it seems that some of the things they are doing is/are triggering this error and I'd like to check if my fix solved the issue

**Comment by franck-boullier** [comment]
BZ database schema has been updated to v3.27 in the DEV/Staging and in the DEMO environment.

We will need to see if there is no unexpected issue before we can roll this out in the PROD environment.

Waiting for the results of the tests by @PicoCreator and report from @kaihendry: is the table `user_group_map` still truncated after DB schema has been updated to v3.27?

**Comment by franck-boullier** [comment]
First manual tests in the demo environment look OK:
- it is possible to create a unit
- it is possible to invite a new user to an existing case

We need more extensive tests before rolling out to prod though...

**Comment by kaihendry** [comment]
When I roll back again to 2AM, obviously the old schema will be in place. Can you document how the schema was updated please @franck-boullier

**Comment by franck-boullier** [comment]
We will keep this issue open in case the issue reappears.
We will monitor the situation for the next 15 days.
If the table `user_group_map` is not truncated after 15 days, then we will close this issue
